docs/static-evaluation.md

```diff
@@ -0,0 +1,388 @@
+* Be constructed with variables derived from a parent static context
+  - This is primarlly for passing values down the module call stack, while maintaining references
+
+There are three potential paths in implementing a static evaluator:
+* Build a custom streamlined (?) solution for this specific problem and it's current use cases
+  - This approach was taken in the prototypes
+  - Can be flexible during development
+  - Does not break other packages
+  - Tests must be written from scratch
+* Re-use existing components of the tofu and lang packages with new plumbing
+  - Can build on top of existing tested logic
+  - Somewhat flexible as components can be swapped out as needed
+  - May require refactoring existing componets we wish to use
+  - May accidentally break other packages due to poor existing testing
+* Re-use current evaluator/scope constructs in tofu and lang packages
+  - Would require re-designing these components to function in either mode
+  - Would come with most of the core logic already implemented
+  - High likelyhood of breaking other packages due to poor existing testing
+  - Would likely require some ergonomic gymnastics depending on scale of refactoring
+
+This will need to be investigated and roughly prototyped, but all solutions should fit a similar enough interface to not block development of dependent tasks. We should design the interface first, based on the requirements of the initial prototype.
+
+
+## Solutions
+### Module Sources
+Module sources must be known at init time as they are downloaded and collated into .terraform/modules. This can be implemented by inspecting the source hcl.Expression using the static evaluator scoped to the current module.
+
+This is relatively straight forward once the core is implemented, but will require some more in-depth changes to support for_each/count later on.
+
+Without module pre-expansion / support for for_each/count, the process would look like:
+* Create a SourceExpression field in config.ModuleCall and don't set the "config.ModuleCall.Source" field initially
+* Use the static context available during NewModule constructor to evaluate all of the config.ModuleCall source fields and check for bad references and other errors.
+
+### Provider Iteration
+
+In [#300](https://github.com/opentofu/opentofu/issues/300), users describe how supporting for_each for configuring providers will allow much DRYer and simpler configurations.
+
+> [!Note]
+> Please read [Provider References](provider-references.md) before diving into this section!
+
+#### Proposed Changes
+
+The first change is to support static variables in the provider config block.  This can then be extended to support for_each/count and be expanded at the end of the `config.NewModule()` function, similar to how module.ProviderLocalNames is generated.  This piece is fairly straightforward and can be done relatively easily.
```

Comment:
Can you please provide an example for this case? I think I got it, but it'll reassure me if I'll see the TF code describing this scenario.